KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h

```diff
@@ -0,0 +1,449 @@
+          is_non_parallel_first = true;
+          is_non_parallel_occur = true;
+        }
+        else {
+          is_non_parallel_first = false;
+        }
+      }
+      else if (height_target < height_source) {
+        dmin_dir = -support_heights[i] + height_source;
+        dmax_dir = support_heights[i + num_directions/2] + height_source;
+
+        rmin_dir = -height_target + height_source;
+        rmax_dir = rmin_dir;
+
+        if ( is_non_parallel_occur == false ) {
+          is_non_parallel_first = true;
+          is_non_parallel_occur = true;
+        }
+        else {
+          is_non_parallel_first = false;
+        }
+      }
+      else {
+        // height_source = height_target, i.e. the ray is parallel to the i-th direction.
+        // This case has already been checked previously, so just inherit tmin and tmax.
+        dmin_dir = dmin;
+        dmax_dir = dmax;
+        rmin_dir = rmin;
+        rmax_dir = rmax;
+
+        if (num_parallels == i) {
+          is_non_parallel_occur = false;
+          num_parallels += 1;
+        }
+      }
+
+      if (i == 0 || is_non_parallel_first == true) {
+        dmin = dmin_dir;
+        dmax = dmax_dir;
+        rmin = rmin_dir;
+        rmax = rmax_dir;
+      }
+      else {
+        if ( is_non_parallel_occur == true && is_non_parallel_first == false ) {
+          // no overlapping if tmin > tmax_dir or tmax < tmin_dir
+          if ( dmin*rmax_dir > rmin*dmax_dir || dmax*rmin_dir < rmax*dmin_dir ) return false;
+
+          // if tmin < tmin_dir, tmin = tmin_dir (narrowing the scope of t)
+          if ( dmin*rmin_dir < rmin*dmin_dir ) dmin = dmin_dir, rmin = rmin_dir;
+
+          // if tmax > tmax_dir, tmax = tmax_dir (narrowing the scope of t)
+          if ( dmax*rmax_dir > rmax*dmax_dir ) dmax = dmax_dir, rmax = rmax_dir;
+        }
+      }
+
+    }
+
+    return is_overlap;
+  }
+
+  template<typename GeomTraits, unsigned int N>
+  bool KDOP_kdop<GeomTraits, N>::do_overlap_sphere(const Array_height& support_heights,
+                                                   const FT& squared_radius) const
+  {
+    bool is_overlap = true;
+
+    const Array_height& support_heights_point = this->support_heights();
+
+    typedef typename Coercion_traits<double, FT>::Type CFT;
+
+    CFT distance = 0.;
+
+    for (int i = 0; i < 3; ++i) {
+      CFT d = 0.;
+      if ( support_heights_point[i] > support_heights[i] ) {
+        d = support_heights_point[i] - support_heights[i]; // distance between the point and the slab
+        if (d * d > squared_radius) return false;
+        distance += d * d; // compute Cartesian distance
+      }
+      else if ( -support_heights_point[i] > support_heights[i + num_directions/2] ) {
+        d = -support_heights_point[i] - support_heights[i + num_directions/2];
+        if (d * d > squared_radius) return false;
+        distance += d * d;
+      }
+    }
+
+    if (distance > squared_radius) return false;
```

Comment:
Must this test not also be done after the next for loop?
And why is it not done after the  `distance +=`  with distance IN the two loops?

Reply:
For a k-dop, it is not straightforward to compute the distance between a point and a polytope (considering the polytope is the intersections of k half-planes), so this distance is not used for k>6 and it is simplified to compare only the support heights in each direction. This may leave some regions with false positives, but those regions are very small. The exact computation of these regions may not have benefit in reducing the time.

The 'distance' is the distance between the point and the bbox. We could also put this check 'distance > squared_radius' in the loops right after 'distance += ...'.